pkg/service/backup/worker_manifest.go

```diff
@@ -88,10 +92,60 @@ func (w *worker) createTemporaryManifest(h hostInfo, tokens []int64) ManifestInf
 		c.Size += d.Progress.Size
 	}
 
+	c.ClusterID = w.ClusterID
+	c.NodeID = h.ID
+	c.DC = h.DC
+
+	rack, err := w.Client.HostRack(ctx, h.IP)
+	if err != nil {
+		return ManifestInfoWithContent{}, errors.Wrap(err, "client.HostRack")
+	}
+	c.Rack = rack
+
+	instanceDetails, err := w.manifestInstanceDetails(ctx, h)
+	if err != nil {
+		return ManifestInfoWithContent{}, errors.Wrap(err, "manifest instance details")
+	}
+	c.InstanceDetails = instanceDetails
+
 	return ManifestInfoWithContent{
 		ManifestInfo:             m,
 		ManifestContentWithIndex: c,
+	}, nil
+}
+
+// manifestInstanceDetails collects node/instance specific information that's needed for 1-to-1 restore.
+func (w *worker) manifestInstanceDetails(ctx context.Context, host hostInfo) (InstanceDetails, error) {
+	var result InstanceDetails
+
+	shardCound, err := w.Client.ShardCount(ctx, host.IP)
+	if err != nil {
+		return InstanceDetails{}, errors.Wrap(err, "client.ShardCount")
+	}
+	result.ShardCount = int(shardCound)
+
+	nodeInfo, err := w.Client.NodeInfo(ctx, host.IP)
+	if err != nil {
+		return InstanceDetails{}, errors.Wrap(err, "client.NodeInfo")
 	}
+	result.StorageSize = nodeInfo.StorageSize
+
+	metaSvc, err := cloudmeta.NewCloudMeta(w.Logger)
+	if err != nil {
+		return InstanceDetails{}, errors.Wrap(err, "new cloud meta svc")
+	}
+
+	instanceMeta, err := metaSvc.GetInstanceMetadata(ctx)
+	if err != nil {
+		// Metadata may not be available for several reasons:
+		// 1. running on-premise 2. disabled 3. smth went wrong with metadata server.
+		// As we cannot distiguish between this cases we can only log err and continue with backup.
+		w.Logger.Error(ctx, "Get instance metadata", "err", err)
+	}
+	result.CloudProvider = string(instanceMeta.CloudProvider)
+	result.InstanceType = instanceMeta.InstanceType
+
+	return result, nil
```

Comment:
Unfortunately, all of this code is executed on the SM side.
I guess that querying instance metadata should be done on the agent side?

Reply:
oh, I missed that, you're right!

Reply:
Do you think NodeInfo is a good place to [extend](https://github.com/scylladb/scylla-manager/blob/164ff2ce065cd106253fed40d2f6ffe883e73cc6/v3/swagger/gen/agent/models/node_info.go#L18) with InstanceType, CloudProvider information? Or it's better to have a separate call for them?

Reply:
I wouldn't mix them if it meant that we always try to query multiple providers when fetching NodeInfo, as it's used in multiple places in the code.
But I guess that it's safe to assume that the instance type won't change at runtime, so we can cache this value in agent memory and query it only once per agent re-start. The problem could be to distinguish cases of getting timeout on querying instance type vs querying it on on-prem.
Another approach could be to extend NodeInfo API call to include optional query param specifying that the instance details should also be included into the node info, but that's not that different from writing a separate endpoint for them.

Reply:
I think I'll go with separate call just for the metadata